sdk/src/main/scala/com/horizen/certificatesubmitter/strategies/WithKeyRotationCircuitStrategy.scala

```diff
@@ -106,50 +104,61 @@ class WithKeyRotationStrategy(settings: SidechainSettings, params: NetworkParams
     val endEpochCumCommTreeHash: Array[Byte] = lastMainchainBlockCumulativeCommTreeHashForWithdrawalEpochNumber(history, referencedWithdrawalEpochNumber)
     val sidechainId = params.sidechainId
 
-    val actualKeys = getSchnorrKeysSignaturesList(getSchnorrKeysSignaturesListBytes(state, referencedWithdrawalEpochNumber))
+    val actualKeys = CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation
+      .getSchnorrKeysSignaturesList(getSchnorrKeysSignaturesListBytes(state, referencedWithdrawalEpochNumber))
     val customFields = actualKeys.getUpdatedKeysRootHash.serializeFieldElement()
 
-    CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation
+    val message = CryptoLibProvider.thresholdSignatureCircuitWithKeyRotation
       .generateMessageToBeSigned(withdrawalRequests.asJava, sidechainId, referencedWithdrawalEpochNumber,
         endEpochCumCommTreeHash, btrFee, ftMinAmount, util.Arrays.asList(customFields))
+
+    actualKeys.getSigningKeys.foreach(_.freePublicKey())
+    actualKeys.getMasterKeys.foreach(_.freePublicKey())
+    actualKeys.getUpdatedSigningKeys.foreach(_.freePublicKey())
+    actualKeys.getUpdatedMasterKeys.foreach(_.freePublicKey())
+    actualKeys.getUpdatedSigningKeysSkSignatures.foreach(_.freeSignature())
+    actualKeys.getUpdatedSigningKeysMkSignatures.foreach(_.freeSignature())
+    actualKeys.getUpdatedMasterKeysSkSignatures.foreach(_.freeSignature())
+    actualKeys.getUpdatedMasterKeysMkSignatures.foreach(_.freeSignature())
```

Comment:
it is a MAJOR sign that something is done incorrectly.
No-one except `ThresholdSignatureCircuitWithKeyRotationImplZendoo` class should care about the native objects
@sgonta

Reply:
Done